client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java

```diff
@@ -83,116 +88,176 @@
+                .retryingOn(StatusRuntimeException.class)
+                .throwingOn(Exception.class);
+
         // Create Async RPC client.
-        client = ControllerServiceGrpc.newStub(channelBuilder.build());
+        this.client = ControllerServiceGrpc.newStub(channelBuilder.build());
     }
 
     @Override
     public CompletableFuture<Boolean> createScope(final String scopeName) {
+        Exceptions.checkNotClosed(closed.get(), this);
         long traceId = LoggerHelpers.traceEnter(log, "createScope", scopeName);
 
-        RPCAsyncCallback<CreateScopeStatus> callback = new RPCAsyncCallback<>();
-        client.createScope(ScopeInfo.newBuilder().setScope(scopeName).build(), callback);
-        return callback.getFuture()
-                .thenApply(x -> {
-                    switch (x.getStatus()) {
-                    case FAILURE:
-                        log.warn("Failed to create scope: {}", scopeName);
-                        throw new ControllerFailureException("Failed to create scope: " + scopeName);
-                    case INVALID_SCOPE_NAME:
-                        log.warn("Illegal scope name: {}", scopeName);
-                        throw new IllegalArgumentException("Illegal scope name: " + scopeName);
-                    case SCOPE_EXISTS:
-                        log.warn("Scope already exists: {}", scopeName);
-                        return false;
-                    case SUCCESS:
-                        log.info("Scope created successfully: {}", scopeName);
-                        return true;
-                    case UNRECOGNIZED:
-                    default:
-                        throw new ControllerFailureException("Unknown return status creating scope " + scopeName
-                                                             + " " + x.getStatus());
-                    }
-                }).whenComplete((x, e) -> {
-                    if (e != null) {
-                        log.warn("createScope failed: ", e);
-                    }
-                    LoggerHelpers.traceLeave(log, "createScope", traceId);
-                });
+        final CompletableFuture<CreateScopeStatus> result = this.retryConfig.runAsync(() -> {
+                RPCAsyncCallback<CreateScopeStatus> callback = new RPCAsyncCallback<>();
+                client.createScope(ScopeInfo.newBuilder().setScope(scopeName).build(), callback);
+                return callback.getFuture();
+        }, this.executor);
+        return result.thenApply(x -> {
+                switch (x.getStatus()) {
+                case FAILURE:
+                    log.warn("Failed to create scope: {}", scopeName);
+                    throw new ControllerFailureException("Failed to create scope: " + scopeName);
+                case INVALID_SCOPE_NAME:
+                    log.warn("Illegal scope name: {}", scopeName);
+                    throw new IllegalArgumentException("Illegal scope name: " + scopeName);
+                case SCOPE_EXISTS:
```

Comment:
Reviewing all the APIs for idempotency is a pending task.
Create an issue to track this - https://github.com/pravega/pravega/issues/1701